File: my-code/1056-capacity-to-ship-packages-within-d-days/solution.cpp
class Solution {
public:
     int check(int cap,vector<int> &weights){
        int currsum = 0;
        int totalday = 0;
         int n = weights.size();
         for(int i=0;i<n;i++){
                if(weights[i]>cap){
                    return INT_MAX;
                }
               currsum+=weights[i];
               if(currsum==cap){
                   totalday++;
                   currsum=0;
               }
              if(currsum>cap){
                   totalday++;
                  currsum=0;
                  currsum = weights[i];
              }
          }
         if(currsum!=0){
               totalday++;
          }
         return totalday;

    }
    int shipWithinDays(vector<int>& weights, int days) {
        
        int low=1;
        int high =INT_MAX;
        int result = -1;
        while(low<high){
                 int mid = low+(high-low)/2;   //mycapacity:
                 if(check(mid,weights)<=days){
                          high = mid;
                          result = mid;
                 }
                 else{
                        low = mid+1;
                 }
        }
        
        return result;
    }
};

File: my-code/1056-capacity-to-ship-packages-within-d-days/solution.java
class Solution {
   
     public static int solve(int cap,int arr[])
    {
        int count=1;
        int sum=0;
        for(int a:arr)
        {
            if(a>cap)
            return Integer.MAX_VALUE;
            
            if(sum+a<=cap)
            {
                sum+=a;
            }
            else
            {
                count++;
                sum=a;
            }
        }
        return count;
    }


    public int shipWithinDays(int[] arr, int D) {
         
        int start=1, end=(int)1e9;
        while(start<end)
        {
            int mid=start+(end-start)/2;
            int days=solve(mid,arr);
            if(days<=D)
            end=mid;
            else
            start=mid+1;
        }
        return start;

    }
}
